File: innerlibs1/ColorExtensions.vb
Imports System.Drawing
Imports System.Runtime.CompilerServices
Imports System.Text.RegularExpressions

''' <summary>
''' Modulo de Conversão de Cores
''' </summary>
''' <remarks></remarks>
Public Module ColorExtensions


    ''' <summary>
    ''' Retorna a <see cref="ConsoleColor"/> mais proxima de uma <see cref="Color"/>
    ''' </summary>
    ''' <param name="Color"></param>
    ''' <returns></returns>
    <Extension> Public Function ToConsoleColor(ByVal Color As System.Drawing.Color) As System.ConsoleColor
        Dim index As Integer = If((Color.R > 128 Or Color.G > 128 Or Color.B > 128), 8, 0)
        index = index Or If((Color.R > 64), 4, 0)
        index = index Or If((Color.G > 64), 2, 0)
        index = index Or If((Color.B > 64), 1, 0)
        Return CType(index, System.ConsoleColor)
    End Function

    ''' <summary>
    ''' Retorna a <see cref="Color"/> a partir de uma <see cref="ConsoleColor"/>
    ''' </summary>
    ''' <param name="Color"></param>
    ''' <returns></returns>
    <Extension> Public Function ToColor(ByVal Color As System.ConsoleColor) As HSVColor
        Dim cColors As Integer() = {&H0, &H80, &H8000, &H8080, &H800000, &H800080, &H808000, &HC0C0C0, &H808080, &HFF, &HFF00, &HFFFF, &HFF0000, &HFF00FF, &HFFFF00, &HFFFFFF}
        Return New HSVColor(cColors(CInt(Color))) With {.Alpha = 255}
    End Function

    <Extension()> Public Function ToHSVColorList(ColorList As IEnumerable(Of Color)) As IEnumerable(Of HSVColor)
        Return ColorList?.Select(Function(x) New HSVColor(x))
    End Function
    Public Function GrayscalePallete(Amount As Integer) As IEnumerable(Of HSVColor)
        Return MonochromaticPallete(Color.White, Amount)
    End Function

    ''' <summary>
    ''' Gera uma paleta de cores monocromatica com <paramref name="Amount"/> amostras a partir de uma <paramref name="Color"/> base.
    ''' </summary>
    ''' <param name="Color"></param>
    ''' <param name="Amount"></param>
    ''' <returns></returns>
    ''' <remarks>A distancia entre as cores será maior se a quantidade de amostras for pequena</remarks>
    Public Function MonochromaticPallete(Color As Color, Amount As Integer) As IEnumerable(Of HSVColor)

        Dim t = New RuleOfThree(Amount, 100, 1, Nothing)

        Dim Percent = t.UnknowValue?.ToSingle()

        Color = Color.White.MergeWith(Color)

        Dim l As New List(Of Color)
        For index = 1 To Amount
            Color = Color.MakeDarker(Percent)
            l.Add(Color)
        Next
        Return l.ToHSVColorList
    End Function

    ''' <summary>
    ''' Retorna  a cor negativa de uma cor
    ''' </summary>
    ''' <param name="TheColor">Cor</param>
    ''' <returns></returns>
    <Extension()>
    Public Function GetNegativeColor(TheColor As Color) As Color
        Return Color.FromArgb(255 - TheColor.R, 255 - TheColor.G, 255 - TheColor.B)
    End Function

    ''' <summary>
    ''' Retorna uma cor de contraste baseado na iluminacao da primeira cor: Uma cor clara se a primeira for escura. Uma cor escura se a primeira for clara
    ''' </summary>
    ''' <param name="TheColor">Primeira cor</param>
    ''' <param name="Percent">Grau de mesclagem da cor escura ou clara</param>
    ''' <returns>Uma cor clara se a primeira cor for escura, uma cor escura se a primeira for clara</returns>
    <Extension()>
    Public Function GetContrastColor(TheColor As Color, Optional Percent As Single = 70) As Color
        Dim a As Double = 1 - (0.299 * TheColor.R + 0.587 * TheColor.G + 0.114 * TheColor.B) / 255
        Dim d = If(a < 0.5, 0, 255)
        Return TheColor.MergeWith(Color.FromArgb(d, d, d), Percent)
    End Function

    ''' <summary>
    ''' Verifica se uma cor é escura
    ''' </summary>
    ''' <param name="TheColor">Cor</param>
    ''' <returns></returns>
    <Extension>
    Public Function IsDark(TheColor As Color) As Boolean
        Return New HSVColor(TheColor).IsDark
    End Function

    ''' <summary>
    ''' Verifica se uma clor é clara
    ''' </summary>
    ''' <param name="TheColor">Cor</param>
    ''' <returns></returns>
    <Extension>
    Public Function IsLight(TheColor As Color) As Boolean
        Return Not TheColor.IsDark()
    End Function

    ''' <summary>
    ''' Mescla duas cores a partir de uma porcentagem
    ''' </summary>
    ''' <param name="TheColor">Cor principal</param>
    ''' <param name="AnotherColor">Cor de mesclagem</param>
    ''' <param name="percent">Porcentagem de mescla</param>
    ''' <returns></returns>
    <Extension>
    Public Function MergeWith(TheColor As Color, AnotherColor As Color, Optional Percent As Single = 50) As Color
        Return TheColor.Lerp(AnotherColor, Percent / 100)
    End Function

    ''' <summary>
    ''' Escurece a cor mesclando ela com preto
    ''' </summary>
    ''' <param name="TheColor">Cor</param>
    ''' <param name="percent">porcentagem de mesclagem</param>
    ''' <returns></returns>
    <Extension>
    Public Function MakeDarker(TheColor As Color, Optional Percent As Single = 50) As Color
        Return TheColor.MergeWith(Color.Black, Percent)
    End Function

    ''' <summary>
    ''' Clareia a cor mistuando ela com branco
    ''' </summary>
    ''' <param name="TheColor">Cor</param>
    ''' <param name="percent">Porcentagem de mesclagem</param>
    ''' <returns></returns>
    <Extension>
    Public Function MakeLighter(TheColor As Color, Optional Percent As Single = 50) As Color
        Return TheColor.MergeWith(Color.White, Percent)
    End Function

    ''' <summary>
    ''' Mescla duas cores usando Lerp
    ''' </summary>
    ''' <param name="FromColor">Cor</param>
    ''' <param name="ToColor">Outra cor</param>
    ''' <param name="amount">Indice de mesclagem</param>
    ''' <returns></returns>
    <Extension>
    Public Function Lerp(FromColor As Color, ToColor As Color, Amount As Single) As Color
        ' start colours as lerp-able floats
        Dim sr As Single = FromColor.R, sg As Single = FromColor.G, sb As Single = FromColor.B
        ' end colours as lerp-able floats
        Dim er As Single = ToColor.R, eg As Single = ToColor.G, eb As Single = ToColor.B
        ' lerp the colours to get the difference
        Dim r As Byte = CByte(sr.Lerp(er, Amount)), g As Byte = CByte(sg.Lerp(eg, Amount)), b As Byte = CByte(sb.Lerp(eb, Amount))
        ' return the new colour
        Return Color.FromArgb(r, g, b)
    End Function

    ''' <summary>
    ''' Converte uma cor de sistema para hexadecimal
    ''' </summary>
    ''' <param name="Color">Cor do sistema</param>
    ''' <param name="Hash">parametro indicando se a cor deve ser retornada com ou sem hashsign (#)</param>
    ''' <returns>string contendo o hexadecimal da cor</returns>

    <Extension()>
    Public Function ToHexadecimal(Color As System.Drawing.Color, Optional Hash As Boolean = True) As String
        Return (Color.R.ToString("X2") & Color.G.ToString("X2") & Color.B.ToString("X2")).PrependIf("#", Hash)
    End Function

    ''' <summary>
    ''' Converte uma cor de sistema para CSS RGB
    ''' </summary>
    ''' <param name="Color">Cor do sistema</param>
    ''' <returns>String contendo a cor em RGB</returns>

    <Extension()>
    Public Function ToCssRGB(Color As System.Drawing.Color) As String
        Return "rgb(" & Color.R.ToString() & "," & Color.G.ToString() & "," & Color.B.ToString() & ")"
    End Function

    <Extension()>
    Public Function ToCssRGBA(Color As System.Drawing.Color) As String
        Return "rgba(" & Color.R.ToString() & "," & Color.G.ToString() & "," & Color.B.ToString() & "," & Color.A.ToString() & ")"
    End Function

    <Extension()> Public Function IsHexaDecimalColor(ByVal Text As String) As Boolean
        Text = Text.RemoveFirstEqual("#")
        Dim myRegex As Regex = New Regex("^[a-fA-F0-9]+$")
        Return Text.IsNotBlank AndAlso myRegex.IsMatch(Text)
    End Function




    ''' <summary>
    ''' Gera uma cor a partir de uma palavra
    ''' </summary>
    ''' <param name="Text">Pode ser um texto em branco (Cor aleatória), uma <see cref="KnownColor"/> (retorna aquela cor exata) ou uma palavra qualquer (gera proceduralmente uma cor)</param>
    ''' <returns></returns>
    <Extension> Public Function ToColor(Text As String) As Color

        If Text.IsBlank() Then
            Return Color.Transparent
        End If

        If Text.IsIn("random", "rand") Then
            Return RandomColor()
        End If



        If Text.IsNumber Then
            Return Color.FromArgb(Text.ToInteger())
        End If

        If Text.IsIn(KnowColors.Select(Function(x) x.Name), StringComparer.InvariantCultureIgnoreCase) Then
            Return KnowColors.FirstOrDefault(Function(x) x.Name.ToLower() = Text.ToLower())
        End If

        If Text.IsHexaDecimalColor Then
            Return ColorTranslator.FromHtml("#" & Text.RemoveFirstEqual("#").IfBlank("000000"))
        End If

        Dim coresInt = Text.GetWords.Select(Function(p) p.ToCharArray().Sum(Function(a) AscW(a) ^ 2 * p.Length)).Sum()

        Return Color.FromArgb(255, Color.FromArgb(coresInt))

    End Function

    ''' <summary>
    ''' Gera uma cor aleatória misturandoo ou não os canais RGB
    ''' </summary>
    ''' <param name="Red">-1 para Random ou de 0 a 255 para especificar o valor</param>
    ''' <param name="Green">-1 para Random ou de 0 a 255 para especificar o valor</param>
    ''' <param name="Blue">-1 para Random ou de 0 a 255 para especificar o valor</param>
    ''' <returns></returns>
    Public Function RandomColor(Optional Red As Integer = -1, Optional Green As Integer = -1, Optional Blue As Integer = -1, Optional Alpha As Integer = 255) As Color
        Red = If(Red < 0, RandomNumber(0, 255), Red).LimitRange(Of Integer)(0, 255)
        Green = If(Green < 0, RandomNumber(0, 255), Green).LimitRange(Of Integer)(0, 255)
        Blue = If(Blue < 0, RandomNumber(0, 255), Blue).LimitRange(Of Integer)(0, 255)
        Alpha = Alpha.LimitRange(Of Integer)(0, 255)
        Return Color.FromArgb(Alpha, Red, Green, Blue)
    End Function

    ''' <summary>
    ''' Lista com todas as <see cref="KnownColor"/> convertidas em <see cref="System.Drawing.Color"/> (Igonora as systemcolors)
    ''' </summary>
    ''' <returns></returns>
    Public ReadOnly Property KnowColors As IEnumerable(Of Color)
        Get
            Return [Enum].GetValues(GetType(KnownColor)).Cast(Of KnownColor).Select(Function(x) Color.FromKnownColor(x)).Where(Function(x) x.IsSystemColor = False)
        End Get
    End Property

    ''' <summary>
    ''' Retorna uma <see cref="KnownColor"/> mais proxima de outra cor
    ''' </summary>
    ''' <param name="Color"></param>
    ''' <returns></returns>
    <Extension()> Public Function GetClosestKnowColor(Color As Color) As Color
        Dim closest_distance As Double = Double.MaxValue
        Dim closest As Color = Color.White
        For Each kc In KnowColors
            Dim d = EuclideanDistance(Color, kc)
            If d < closest_distance Then
                closest_distance = d
                closest = kc
            End If
        Next
        Return closest
    End Function

    ''' <summary>
    ''' Retorna a distancia euclideana
    ''' </summary>
    ''' <param name="Color"></param>
    ''' <param name="OtherColor"></param>
    ''' <returns></returns>
    <Extension> Public Function EuclideanDistance(Color As Color, OtherColor As Color) As Double
        Dim r_dist_sqrd As Double = Math.Pow(CDbl(Color.R) - CDbl(OtherColor.R), 2)
        Dim g_dist_sqrd As Double = Math.Pow(CDbl(Color.G) - CDbl(OtherColor.G), 2)
        Dim b_dist_sqrd As Double = Math.Pow(CDbl(Color.B) - CDbl(OtherColor.B), 2)
        Return Math.Sqrt(r_dist_sqrd + g_dist_sqrd + b_dist_sqrd)
    End Function

    ''' <summary>
    ''' Retorna o nome comum mais proximo a esta cor
    ''' </summary>
    ''' <param name="Color"></param>
    ''' <returns></returns>
    <Extension()> Public Function GetClosestColorName(Color As Color) As String
        Return Color.GetClosestKnowColor().Name
    End Function

    ''' <summary>
    ''' Retorna o nome da cor
    ''' </summary>
    ''' <param name="Color"></param>
    ''' <returns></returns>
    <Extension()> Public Function GetColorName(Color As Color) As String
        For Each namedColor In KnowColors
            If Color.CompareARGB(namedColor) Then Return namedColor.Name
        Next
        Return Color.Name
    End Function

    ''' <summary>
    ''' Verifica se uma cor é legivel sobre outra
    ''' </summary>
    ''' <param name="Color"></param>
    ''' <param name="BackgroundColor"></param>
    ''' <param name="Size"></param>
    ''' <returns></returns>
    <Extension()> Public Function IsReadable(Color As Color, BackgroundColor As Color, Optional Size As Integer = 10) As Boolean
        If Color.A = 0 Then Return False
        If BackgroundColor.A = 0 Then Return True
        Dim diff = BackgroundColor.R * 0.299 + BackgroundColor.G * 0.587 + BackgroundColor.B * 0.114 - Color.R * 0.299 - Color.G * 0.587 - Color.B * 0.114
        Return Not ((diff < (1.5 + 141.162 * Math.Pow(0.975, Size)))) AndAlso (diff > (-0.5 - 154.709 * Math.Pow(0.99, Size)))
    End Function

End Module

File: InnerLibs/ColorExtensions.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text.RegularExpressions;

namespace InnerLibs
{
    /// <summary>
    /// Modulo de Conversão de Cores
    /// </summary>
    /// <remarks></remarks>
    public static class ColorExtensions
    {
        #region Public Properties

        /// <summary>
        /// Retorna uma lista com todas as <see cref="KnowColor"/> convertidas em <see cref="System.Drawing.Color"/>
        /// </summary>
        public static IEnumerable<Color> KnowColors => Ext.GetEnumValues<KnownColor>().Select(x => Color.FromKnownColor(x));

        #endregion Public Properties

        #region Public Methods

        /// <summary>
        /// Procura uma cor na tabela de cores <see cref="HSVColor.NamedColors"/>
        /// </summary>
        /// <param name="Text"></param>
        /// <returns></returns>
        public static HSVColor FindColor(this string Text) => HSVColor.NamedColors
               .FirstOrDefault(x => x.Name.ToLowerInvariant().Replace("grey", "gray").RemoveAny(PredefinedArrays.PasswordSpecialChars.Union(new[] { " " }).ToArray()) == Text.ToLowerInvariant().Replace("grey", "gray").RemoveAny(PredefinedArrays.PasswordSpecialChars.Union(new[] { " " }).ToArray()));

        /// <summary>
        /// Retorna o nome comum mais proximo a esta cor
        /// </summary>
        /// <param name="Color"></param>
        /// <returns></returns>
        public static string GetClosestColorName(this Color Color) => Color.GetClosestKnowColor().Name;

        /// <summary>
        /// Retorna uma cor conhecida mais proxima de outra cor
        /// </summary>
        /// <param name="Color"></param>
        /// <returns></returns>
        public static Color GetClosestKnowColor(this Color Color)
        {
            double closest_distance = double.MaxValue;
            var closest = Color.White;
            foreach (var kc in KnowColors)
            {
                // Calculate Euclidean Distance
                double d = new HSVColor(kc).GetEuclideanDistance(Color);
                if (d < closest_distance)
                {
                    closest_distance = d;
                    closest = kc;
                }
            }

            return closest;
        }

        /// <summary>
        /// Retorna o nome da cor
        /// </summary>
        /// <param name="Color"></param>
        /// <returns></returns>
        public static string GetColorName(this Color Color)
        {
            foreach (var namedColor in HSVColor.NamedColors) if (namedColor.ARGB == Color.ToArgb()) return namedColor.Name;
            return Color.Name;
        }

        /// <summary>
        /// Retorna uma cor de contraste baseado na iluminacao da primeira cor: Uma cor clara se a
        /// primeira for escura. Uma cor escura se a primeira for clara
        /// </summary>
        /// <param name="TheColor">Primeira cor</param>
        /// <param name="Percent">Grau de mesclagem da cor escura ou clara</param>
        /// <returns>
        /// Uma cor clara se a primeira cor for escura, uma cor escura se a primeira for clara
        /// </returns>
        public static Color GetContrastColor(this Color TheColor, float Percent = 70f)
        {
            double a = 1d - (0.299d * TheColor.R + 0.587d * TheColor.G + 0.114d * TheColor.B) / 255d;
            int d = a < 0.5d ? 0 : 255;
            return TheColor.MergeWith(Color.FromArgb(d, d, d), Percent);
        }

        /// <summary>
        /// Retorna a cor negativa de uma cor
        /// </summary>
        /// <param name="TheColor">Cor</param>
        /// <returns></returns>
        public static Color GetNegativeColor(this Color TheColor) => Color.FromArgb(255 - TheColor.R, 255 - TheColor.G, 255 - TheColor.B);

        public static IEnumerable<HSVColor> GrayscalePallette(int Amount) => MonochromaticPallette(Color.White, Amount);

        /// <summary>
        /// Verifica se uma cor é escura
        /// </summary>
        /// <param name="TheColor">Cor</param>
        /// <returns></returns>
        public static bool IsDark(this Color TheColor) => new HSVColor(TheColor).IsDark();

        public static bool IsHexaDecimalColor(this string Text)
        {
            Text = Text.TrimFirstEqual("#");
            var myRegex = new Regex("^[a-fA-F0-9]+$");
            return Text.IsNotBlank() && myRegex.IsMatch(Text);
        }

        /// <summary>
        /// Verifica se uma cor é clara
        /// </summary>
        /// <param name="TheColor">Cor</param>
        /// <returns></returns>
        public static bool IsLight(this Color TheColor) => !TheColor.IsDark();

        /// <summary>
        /// Verifica se uma cor é legivel sobre outra
        /// </summary>
        /// <param name="Color"></param>
        /// <param name="BackgroundColor"></param>
        /// <param name="Size"></param>
        /// <returns></returns>
        public static bool IsReadable(this Color Color, Color BackgroundColor, int Size = 10)
        {
            if (Color.A == 0)
                return false;
            if (BackgroundColor.A == 0)
                return true;
            double diff = BackgroundColor.R * 0.299d + BackgroundColor.G * 0.587d + BackgroundColor.B * 0.114d - Color.R * 0.299d - Color.G * 0.587d - Color.B * 0.114d;
            return !(diff < 1.5d + 141.162d * Math.Pow(0.975d, Size)) && diff > -0.5d - 154.709d * Math.Pow(0.99d, Size);
        }

        /// <summary>
        /// Mescla duas cores usando <see cref="Lerp"/>
        /// </summary>
        /// <param name="FromColor">Cor</param>
        /// <param name="ToColor">Outra cor</param>
        /// <param name="amount">Indice de mesclagem</param>
        /// <returns></returns>
        public static Color Lerp(this Color FromColor, Color ToColor, float Amount)
        {
            // start colours as lerp-able floats
            float sr = FromColor.R;
            float sg = FromColor.G;
            float sb = FromColor.B;
            // end colours as lerp-able floats
            float er = ToColor.R;
            float eg = ToColor.G;
            float eb = ToColor.B;
            // lerp the colours to get the difference
            byte r = (byte)Math.Round(sr.Lerp(er, Amount));
            byte g = (byte)Math.Round(sg.Lerp(eg, Amount));
            byte b = (byte)Math.Round(sb.Lerp(eb, Amount));
            // return the new colour
            return Color.FromArgb(r, g, b);
        }

        /// <summary>
        /// Escurece a cor mesclando ela com preto
        /// </summary>
        /// <param name="TheColor">Cor</param>
        /// <param name="percent">porcentagem de mesclagem</param>
        /// <returns></returns>
        public static Color MakeDarker(this Color TheColor, float Percent = 50f) => TheColor.MergeWith(Color.Black, Percent);

        /// <summary>
        /// Clareia a cor misturando ela com branco
        /// </summary>
        /// <param name="TheColor">Cor</param>
        /// <param name="percent">Porcentagem de mesclagem</param>
        /// <returns></returns>
        public static Color MakeLighter(this Color TheColor, float Percent = 50f) => TheColor.MergeWith(Color.White, Percent);

        /// <summary>
        /// Mescla duas cores a partir de uma porcentagem
        /// </summary>
        /// <param name="TheColor">Cor principal</param>
        /// <param name="AnotherColor">Cor de mesclagem</param>
        /// <param name="percent">Porcentagem de mescla</param>
        /// <returns></returns>
        public static Color MergeWith(this Color TheColor, Color AnotherColor, float Percent = 50f) => TheColor.Lerp(AnotherColor, Percent / 100f);

        /// <summary>
        /// Gera uma paleta de cores monocromatica com <paramref name="Amount"/> amostras a partir
        /// de uma <paramref name="Color"/> base.
        /// </summary>
        /// <param name="Color"></param>
        /// <param name="Amount"></param>
        /// <returns></returns>
        /// <remarks>A distancia entre as cores será maior se a quantidade de amostras for pequena</remarks>
        public static IEnumerable<HSVColor> MonochromaticPallette(Color Color, int Amount)
        {
            var t = new RuleOfThree<int>(Amount, 100, 1, default);
            var Percent = t.UnknownValue?.ToSingle();
            Color = Color.White.MergeWith(Color);
            var l = new List<Color>();
            for (int index = 1, loopTo = Amount; index <= loopTo; index++)
            {
                Color = Color.MakeDarker((float)Percent);
                l.Add(Color);
            }

            return l.ToHSVColorList();
        }

        /// <summary>
        /// Gera uma cor aleatória misturando ou não os canais RGB
        /// </summary>
        /// <param name="Red">-1 para Random ou de 0 a 255 para especificar o valor</param>
        /// <param name="Green">-1 para Random ou de 0 a 255 para especificar o valor</param>
        /// <param name="Blue">-1 para Random ou de 0 a 255 para especificar o valor</param>
        /// <returns></returns>
        public static Color RandomColor(int Red = -1, int Green = -1, int Blue = -1, int Alpha = 255)
        {
            Red = Red.SetMinValue(-1);
            Green = Green.SetMinValue(-1);
            Blue = Blue.SetMinValue(-1);

            Red = (Red < 0 ? Ext.RandomNumber(0, 255) : Red).LimitRange<int>(0, 255);
            Green = (Green < 0 ? Ext.RandomNumber(0, 255) : Green).LimitRange<int>(0, 255);
            Blue = (Blue < 0 ? Ext.RandomNumber(0, 255) : Blue).LimitRange<int>(0, 255);
            Alpha = Alpha.LimitRange<int>(0, 255);
            return Color.FromArgb(Alpha, Red, Green, Blue);
        }

        /// <summary>
        /// Retorna a <see cref="Color"/> a partir de uma <see cref="ConsoleColor"/>
        /// </summary>
        /// <param name="Color"></param>
        /// <returns></returns>
        public static HSVColor ToColor(this ConsoleColor Color) => new HSVColor(new[] { 0x0, 0x80, 0x8000, 0x8080, 0x800000, 0x800080, 0x808000, 0xC0C0C0, 0x808080, 0xFF, 0xFF00, 0xFFFF, 0xFF0000, 0xFF00FF, 0xFFFF00, 0xFFFFFF }[(int)Color]) { Alpha = 255 };

        /// <summary>
        /// Gera uma cor a partir de uma palavra
        /// </summary>
        /// <param name="Text">
        /// Pode ser um texto em branco (Transparent), uma <see cref="NamedColors"/> (retorna aquela
        /// cor exata), uma palavra qualquer (gera proceduralmente uma cor) ou uma expressão de cor
        /// (Red+Blue, Red-Blue,Green*Red etc).
        /// </param>
        /// <returns></returns>
        public static Color ToColor(this string Text)
        {
            if (Text.IsBlank()) return Color.Transparent;

            if (Text == "random") return RandomColor();

            if (Text.IsNumber()) return Color.FromArgb(Text.ToInt());

            if (Text.IsHexaDecimalColor()) return ColorTranslator.FromHtml("#" + Text.TrimFirstEqual("#"));

            var maybecolor = FindColor(Text);
            if (maybecolor != null)
            {
                return maybecolor.ToDrawingColor();
            }

            if (Text.Contains("+"))
            {
                var various = Text.Split("+");

                if (various.Any())
                {
                    return various.Select(x => new HSVColor(x.Trim())).Aggregate((a, b) => a + b);
                }
            }

            if (Text.Contains("-"))
            {
                var various = Text.Split("-");
                if (various.Any())
                {
                    return various.Select(x => new HSVColor(x.Trim())).Aggregate((a, b) => a - b);
                }
            }
            if (Text.Contains("*"))
            {
                var various = Text.Split("*");

                if (various.Any())
                {
                    return various.Select(x => new HSVColor(x.Trim())).Aggregate((a, b) => a * b);
                }
            }

            var coresInt = Text.GetWords().Select(p => p.ToCharArray().Sum(a => Math.Pow(a.ToAsc(), 2d) * p.Length)).Sum().RoundInt();
            return Color.FromArgb(255, Color.FromArgb(coresInt));
        }

        /// <summary>
        /// Retorna a <see cref="ConsoleColor"/> mais proxima de uma <see cref="Color"/>
        /// </summary>
        /// <param name="Color"></param>
        /// <returns></returns>
        public static ConsoleColor ToConsoleColor(this Color Color)
        {
            int index = Color.R > 128 | Color.G > 128 | Color.B > 128 ? 8 : 0;
            index |= Color.R > 64 ? 4 : 0;
            index |= Color.G > 64 ? 2 : 0;
            index |= Color.B > 64 ? 1 : 0;
            return (ConsoleColor)index;
        }

        public static string ToCssRGB(this Color Color) => "rgb(" + Color.R.ToString() + "," + Color.G.ToString() + "," + Color.B.ToString() + ")";

        /// <summary>
        /// Converte uma cor de sistema para CSS RGB
        /// </summary>
        /// <param name="Color">Cor do sistema</param>
        /// <returns>String contendo a cor em RGB</returns>
        public static string ToCssRGBA(this Color Color) => "rgba(" + Color.R.ToString() + "," + Color.G.ToString() + "," + Color.B.ToString() + "," + Color.A.ToString() + ")";

        public static string ToHexadecimal(this Color Color, bool Hash = true) => (Color.R.ToString("X2") + Color.G.ToString("X2") + Color.B.ToString("X2")).PrependIf("#", Hash);

        public static IEnumerable<HSVColor> ToHSVColorList(this IEnumerable<Color> ColorList) => ColorList?.Select(x => new HSVColor(x));

        #endregion Public Methods

        /// <summary>
        /// Converte uma cor de sistema para hexadecimal
        /// </summary>
        /// <param name="Color">Cor do sistema</param>
        /// <param name="Hash">
        /// parametro indicando se a cor deve ser retornada com ou sem hashsign (#)
        /// </param>
        /// <returns>string contendo o hexadecimal da cor</returns>
    }
}